usb/esp_tinyusb/tusb_msc_storage.c

```diff
@@ -0,0 +1,359 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// DESCRIPTION:
+// This file contains the code for accessing the storage medium.
+
+#include "esp_log.h"
+#include "esp_err.h"
+#include "esp_check.h"
+#include "esp_vfs_fat.h"
+#include "diskio_impl.h"
+#include "diskio_wl.h"
+#include "wear_levelling.h"
+#include "esp_partition.h"
+#include "driver/sdmmc_host.h"
+#include "diskio_sdmmc.h"
+#include "tusb_msc_storage.h"
+
+static wl_handle_t s_wl_handle = WL_INVALID_HANDLE;
+#if SOC_SDMMC_HOST_SUPPORTED
+static sdmmc_card_t *s_card;
+#endif
+static bool s_fat_mounted;
+static const char *s_base_path;
+
+static const char *TAG = "tinyusb_msc_storage";
+
+static esp_err_t _spiflash_init(const tinyusb_msc_storage_config_t *config)
+{
+    ESP_LOGI(TAG, "Initializing wear levelling");
+
+    const esp_partition_t *data_partition = esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_DATA_FAT, NULL);
+    if (data_partition == NULL) {
+        ESP_LOGE(TAG, "Failed to find FATFS partition. Check the partition table.");
+        return ESP_ERR_NOT_FOUND;
+    }
+
+    return wl_mount(data_partition, &s_wl_handle);
+}
+
+#if SOC_SDMMC_HOST_SUPPORTED
+static esp_err_t _sdmmc_init(const tinyusb_msc_storage_config_t *config)
+{
+    esp_err_t ret = ESP_OK;
+    bool host_init = false;
+
+    ESP_LOGI(TAG, "Initializing SDCard");
+
+    sdmmc_host_t host = config->sdmmc_config->host;
+    sdmmc_slot_config_t slot_config = config->sdmmc_config->slot_config;
+
+    // not using ff_memalloc here, as allocation in internal RAM is preferred
+    s_card = (sdmmc_card_t *)malloc(sizeof(sdmmc_card_t));
+    ESP_GOTO_ON_FALSE(s_card, ESP_ERR_NO_MEM, clean, TAG, "could not allocate new sdmmc_card_t");
+
+    ESP_GOTO_ON_ERROR((*host.init)(), clean, TAG, "Host Config Init fail");
+    host_init = true;
+
+    ESP_GOTO_ON_ERROR(sdmmc_host_init_slot(host.slot, (const sdmmc_slot_config_t *) &slot_config),
+                      clean, TAG, "Host init slot fail");
+
+    while (sdmmc_card_init(&host, s_card)) {
+        ESP_LOGE(TAG, "The detection pin of the slot is disconnected(Insert uSD card). Retrying...");
+        vTaskDelay(pdMS_TO_TICKS(3000));
+    }
+
+    // Card has been initialized, print its properties
+    sdmmc_card_print_info(stdout, s_card);
```

Comment:
Probably unnecessary. Does the user need to know any additional about the SD card? (apart form size, available from `tinyusb_msc_storage_get_sector_count/size`)

Reply:
done.
Removed the printing of this additional information.